src/lib.rs

```diff
@@ -0,0 +1,337 @@
+#![crate_type = "staticlib"]
+
+use libc::{c_char, c_int, c_void, size_t, ssize_t};
+use std::io::{Cursor, Read, Write};
+use std::slice;
+use std::{ffi::CStr, sync::Arc};
+use std::{io::ErrorKind::ConnectionAborted, mem};
+
+use rustls::{ClientConfig, ClientSession, Session};
+
+type CrustlsResult = c_int;
+
+pub const CRUSTLS_OK: c_int = 0;
+pub const CRUSTLS_ERROR: c_int = 1;
+
+/// Create a client_config. Caller owns the memory and must free it with
+/// rustls_client_config_free.
+#[no_mangle]
+pub extern "C" fn rustls_client_config_new() -> *const c_void {
+    let mut config = rustls::ClientConfig::new();
+    config
+        .root_store
+        .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
+    env_logger::init();
+    Arc::into_raw(Arc::new(config)) as *const c_void
+}
+
+/// "Free" a client_config previously returned from rustls_client_config_new.
+/// Since client_config is actually an atomically reference-counted pointer,
+/// extant client_sessions may still hold an internal reference to the
+/// Rust object. However, C code must consider this pointer unusable after
+/// "free"ing it.
+/// Calling with NULL is fine. Must not be called twice with the same value.
+#[no_mangle]
+pub extern "C" fn rustls_client_config_free(config: *const c_void) {
+    unsafe {
+        if let Some(c) = (config as *const ClientConfig).as_ref() {
+            // To free the client_config, we reconstruct the Arc. It should have a refcount of 1,
+            // representing the C code's copy. When it drops, that refcount will go down to 0
+            // and the inner ClientConfig will be dropped.
+            let arc: Arc<ClientConfig> = Arc::from_raw(c);
+            let strong_count = Arc::strong_count(&arc);
+            if strong_count < 1 {
```

Comment:
FWIW, I'm pretty sure this is unreachable -- if it was 0 then the value would already have been freed.

Reply:
Hm, good point. My goal here was to catch a double free. And indeed, I can produce unexpected results by double freeing - but they are unsound. I.e. the Arc gets freed on the first call; on the second call, the Arc (including strong_count) is uninitialized. For instance in a test run it winds up as `94281166761504`. I guess there really isn't anything I can do to guard against unsound usage on the C side and I should just delete this.

Reply:
Thanks for pointing out this unreachable code. Looks like I neglected to delete it in this PR. Followed up in https://github.com/abetterinternet/crustls/pull/32.